<a href="https://colab.research.google.com/github/VivianKingasia/Time-Series-Modeling-Forecasting-Zillow-Real-Estate-Prices/blob/main/Time_Series_Modeling_Forecasting_Zillow_Real_Estate_Prices.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# THE LEAGUE PROJECT
Authors:
 - Keith Maina
 - Vivian Kingasia
 - Ann Maureen
 - Brian Kigen
 - Charity Gakuru
 - Hannah Mutua
 - Mercy Ngila
 - Steve Troy

## Business Understanding

### Introduction
In 2021, the real estate industry in the United States was valued at USD 3.69 trillion and was expected to experience a 5.2% compound interest growth for the period between 2022 and 2030. This potential predicted growth of the industry, coupled with rising population rates in the US create a huge lucrative opportunity potential for real estate investors to make huge profits provided they <b>manage risk</b> and <b>make the right investments</b>.<br>
According to <a href= 'https://www.peoplescapitalgroup.com/average-roi-real-estate/'>People's Capital Group</a>, residential properties have an average annual return of 10.6% and commercial properties have a 9.5% average return.

### Problem Statement
The stakeholder in this project is a real estate investment firm that is looking to construct residential homes in top five locations in the US that would provide a high return on their investment. This project therefore is a time series analysis on a Zillow dataset on various locations around the United States.<br><br>

The project will involve analyzing the house sale prices from 1996 to 2018 to determine the top five locations with the highest Return on Investment (ROI).<br>
The stakeholder is also risk-averse and therefore the project involves recommending locations with low price volatility which can easily be predicted with the model.<br>

In our time series analysis, the metric of success to determine model viability will be RMSE/MSE.

### Project Scope
The primary goal of this project will be to conduct a time series analysis to predict the five best locations to invest in based on ROI.

### Problem Questions
- What are the five best locations to invest in around the US?
- What makes these locations so lucrative?
- Does urbanization affect the prices of houses?
- How long does it take to cash out on the investment?
- How risky <b>(measured as ---)</b> is the investment?
- When are the prices most volatile <b>(measured as frequency of price change in a small timeframe)</b> and where are the locations of the houses with most volatility?
- How much profit can investors potentially make based on our predictions?
- Can we use the information gained from this project to gain clients and if so, how?


### Project Objectives
1. Provide effective real estate investment recommendations to the stakeholder.
2. Increase the real estate investor’s customer base.


### DATA CLEANING

In [1]:
#importing libraries
import pandas as pd 

In [2]:
#loading the dataset
df = pd.read_csv('./data/zillow_data.csv')
df.head()

RegionID  RegionName      City State              Metro CountyName  \
0     84654       60657   Chicago    IL            Chicago       Cook   
1     90668       75070  McKinney    TX  Dallas-Fort Worth     Collin   
2     91982       77494      Katy    TX            Houston     Harris   
3     84616       60614   Chicago    IL            Chicago       Cook   
4     93144       79936   El Paso    TX            El Paso    El Paso   

   SizeRank   1996-04   1996-05   1996-06  ...  2017-07  2017-08  2017-09  \
0         1  334200.0  335400.0  336500.0  ...  1005500  1007500  1007800   
1         2  235700.0  236900.0  236700.0  ...   308000   310000   312500   
2         3  210400.0  212200.0  212200.0  ...   321000   320600   320200   
3         4  498100.0  500900.0  503100.0  ...  1289800  1287700  1287400   
4         5   77300.0   77300.0   77300.0  ...   119100   119400   120000   

   2017-10  2017-11  2017-12  2018-01  2018-02  2018-03  2018-04  
0  1009600  1013300  1018700  1024400  1030700  1033800  1030600  
1   314100   315000   316600   318100   319600   321100   321800  
2   320400   320800   321200   321200   323000   326900   329900  
3  1291500  1296600  1299000  1302700  1306400  1308500  1307000  
4   120300   120300   120300   120300   120500   121000   121500  

[5 rows x 272 columns]

In [3]:
#information about the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14723 entries, 0 to 14722
Columns: 272 entries, RegionID to 2018-04
dtypes: float64(219), int64(49), object(4)
memory usage: 30.6+ MB


In [4]:
#getting the columns in the dataset
df.columns

Index(['RegionID', 'RegionName', 'City', 'State', 'Metro', 'CountyName',
       'SizeRank', '1996-04', '1996-05', '1996-06',
       ...
       '2017-07', '2017-08', '2017-09', '2017-10', '2017-11', '2017-12',
       '2018-01', '2018-02', '2018-03', '2018-04'],
      dtype='object', length=272)

##### The first step in the data cleaning process is checking for validity and dropping any columns if necessary. Renaming of columns will also be done here

In [5]:
#Change the column name from RegionName to zipcode beause the column actually contains the zipcodes
df = df.rename(columns={'RegionName': 'Zipcode'})
df.head()

RegionID  Zipcode      City State              Metro CountyName  SizeRank  \
0     84654    60657   Chicago    IL            Chicago       Cook         1   
1     90668    75070  McKinney    TX  Dallas-Fort Worth     Collin         2   
2     91982    77494      Katy    TX            Houston     Harris         3   
3     84616    60614   Chicago    IL            Chicago       Cook         4   
4     93144    79936   El Paso    TX            El Paso    El Paso         5   

    1996-04   1996-05   1996-06  ...  2017-07  2017-08  2017-09  2017-10  \
0  334200.0  335400.0  336500.0  ...  1005500  1007500  1007800  1009600   
1  235700.0  236900.0  236700.0  ...   308000   310000   312500   314100   
2  210400.0  212200.0  212200.0  ...   321000   320600   320200   320400   
3  498100.0  500900.0  503100.0  ...  1289800  1287700  1287400  1291500   
4   77300.0   77300.0   77300.0  ...   119100   119400   120000   120300   

   2017-11  2017-12  2018-01  2018-02  2018-03  2018-04  
0  1013300  1018700  1024400  1030700  1033800  1030600  
1   315000   316600   318100   319600   321100   321800  
2   320800   321200   321200   323000   326900   329900  
3  1296600  1299000  1302700  1306400  1308500  1307000  
4   120300   120300   120300   120500   121000   121500  

[5 rows x 272 columns]

- The dataset that will be analyzed and modelled needs to be complete. All the missing values must be handled.
- On top of that, the data must be consistent. Any duplicates will be handled where necessary.

In [6]:
#Check for null values 
print(f'The data has {df.isna().sum().sum()} missing values')

The data has 157934 missing values


In [7]:
# checking for duplicates

print(f'The data has {df.duplicated().sum()} duplicates')

The data has 0 duplicates


- This data has a lot of missing values in our dataset. Each feature will be investaged separately so that all the missing values can be handled accordingly.
- The data has no dupicates hence it's consistent.

##### RegionID

In [8]:
#What are the unique values in this feature
df.RegionID.unique

<bound method Series.unique of 0        84654
1        90668
2        91982
3        84616
4        93144
         ...  
14718    58333
14719    59107
14720    75672
14721    93733
14722    95851
Name: RegionID, Length: 14723, dtype: int64>

In [9]:
df['RegionID'].isna().any()

False

There are no missing values in this column and all the values are unique

##### Zipcode

In [10]:
#Check for missing values 
df['Zipcode'].isna().sum()

0

In [13]:
#check for duplicates
df.Zipcode.unique

<bound method Series.unique of 0        60657
1        75070
2        77494
3        60614
4        79936
         ...  
14718     1338
14719     3293
14720    40404
14721    81225
14722    89155
Name: Zipcode, Length: 14723, dtype: int64>

All the zipcodes are unique. Furthermore, it has no missing values

In [14]:
#Theck the shape of the zipcodes
print(df.Zipcode.min())
print(df.Zipcode.max())

1001
99901


In [15]:
#Convert all the zipcodes to strings 
df.Zipcode = df.Zipcode.astype('string')

In [16]:
#The zipcodes need to be 5 digits long, so a zero will be added to the ones that have four digits 
for i in range(len(df)):
    df.Zipcode[i] = df.Zipcode[i].rjust(5, '0')

In [17]:
print(df.Zipcode.min())

01001


##### City

In [18]:
#Check for null values 
df['City'].isna().sum()

0

In [19]:
df.City.nunique()

7554

- There are no null values in the City column
- There are 7554 unique cities in the dataset

#### State

In [20]:
#Check for null values 
df['State'].isna().sum()

0

In [21]:
df.State.nunique()

51

- There are no missing values in the State column.
- There are 51 different states in the dataset.

##### Metro

In [23]:
#Check for null values 
df['Metro'].isna().sum()

1043

There are 1043 missing values in the Metro column. This column won't be dropped but instead the fill the missing values with None.

In [24]:
df.Metro.fillna('None', inplace=True)

In [25]:
df.Metro.value_counts()

None                              1043
New York                           779
Los Angeles-Long Beach-Anaheim     347
Chicago                            325
Philadelphia                       281
                                  ... 
Starkville                           1
Pampa                                1
Indianola                            1
Williston                            1
Vernon                               1
Name: Metro, Length: 702, dtype: int64

All the missing values have been dealt with.

#### CountyName

In [26]:
#Chceck for missing values 
df.CountyName.isna().sum()

0

The column has no missing values.

##### SizeRank

In [28]:
#Chceck for missing values 
df.SizeRank.isna().any()

False

In [29]:
df.SizeRank.value_counts()

2047     1
661      1
12947    1
8849     1
10896    1
        ..
3387     1
1338     1
7481     1
5432     1
2049     1
Name: SizeRank, Length: 14723, dtype: int64

This column has no missing values

Now, we will create a time series by changing the data from wide view to long view, and indexing it by the Date.

In [30]:
def melt_df(df):
    melted = pd.melt(df, id_vars=['RegionID','Zipcode', 'City', 'State', 'Metro', 'CountyName', 'SizeRank'], 
                     var_name='Date')
    melted['Date'] = pd.to_datetime(melted['Date'], infer_datetime_format=True)
    melted = melted.dropna(subset=['value'])
    return melted

In [31]:
melted_df = melt_df(df)

In [32]:
# Make sure the data type of the 'Date' column is datetime
melted_df['Date'] = pd.to_datetime(melted_df['Date'], format='%m/%y')

# Set the 'Date' column as index
melted_df.set_index('Date', inplace=True)

In [33]:
#Confirm the index
melted_df.index

DatetimeIndex(['1996-04-01', '1996-04-01', '1996-04-01', '1996-04-01',
               '1996-04-01', '1996-04-01', '1996-04-01', '1996-04-01',
               '1996-04-01', '1996-04-01',
               ...
               '2018-04-01', '2018-04-01', '2018-04-01', '2018-04-01',
               '2018-04-01', '2018-04-01', '2018-04-01', '2018-04-01',
               '2018-04-01', '2018-04-01'],
              dtype='datetime64[ns]', name='Date', length=3744704, freq=None)

###### Uniformity of the data is important. This will be checked by looking at the data types of the different columns and ensuring they are are correct

In [34]:
melted_df.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 3744704 entries, 1996-04-01 to 2018-04-01
Data columns (total 8 columns):
 #   Column      Dtype  
---  ------      -----  
 0   RegionID    int64  
 1   Zipcode     string 
 2   City        object 
 3   State       object 
 4   Metro       object 
 5   CountyName  object 
 6   SizeRank    int64  
 7   value       float64
dtypes: float64(1), int64(2), object(4), string(1)
memory usage: 257.1+ MB


In [35]:
melted_df.head()

RegionID Zipcode      City State              Metro CountyName  \
Date                                                                         
1996-04-01     84654   60657   Chicago    IL            Chicago       Cook   
1996-04-01     90668   75070  McKinney    TX  Dallas-Fort Worth     Collin   
1996-04-01     91982   77494      Katy    TX            Houston     Harris   
1996-04-01     84616   60614   Chicago    IL            Chicago       Cook   
1996-04-01     93144   79936   El Paso    TX            El Paso    El Paso   

            SizeRank     value  
Date                            
1996-04-01         1  334200.0  
1996-04-01         2  235700.0  
1996-04-01         3  210400.0  
1996-04-01         4  498100.0  
1996-04-01         5   77300.0

In [36]:
melted_df.tail()


RegionID Zipcode                 City State            Metro  \
Date                                                                       
2018-04-01     58333   01338             Ashfield    MA  Greenfield Town   
2018-04-01     59107   03293            Woodstock    NH        Claremont   
2018-04-01     75672   40404                Berea    KY         Richmond   
2018-04-01     93733   81225  Mount Crested Butte    CO             None   
2018-04-01     95851   89155             Mesquite    NV        Las Vegas   

           CountyName  SizeRank     value  
Date                                       
2018-04-01   Franklin     14719  209300.0  
2018-04-01    Grafton     14720  225800.0  
2018-04-01    Madison     14721  133400.0  
2018-04-01   Gunnison     14722  664400.0  
2018-04-01      Clark     14723  357200.0